src/dbnode/encoding/proto/docs/unmarshal.md

```diff
@@ -0,0 +1,50 @@
+It's especially inefficient for Protobuf schemas that are optimized for this package (specifically those that make heavy use of top-level scalar fields where allocating objects on the heap just to wrap primitive types is especially wasteful).
+
+As a result, this package implements a `customFieldUnmarshaler` that accepts a `[]byte` which constitutes a marshaled Protobuf message and exposes methods for unmarshaling the top-level scalar fields (i.e the fields that the encoder can perform custom compression on) in an efficient and reusable manner such that in the general case there are no allocations.
+In addition, it uses the `jhump/protoreflect` library to unmarshal any other fields that it can't unmarshal efficiently (although this has zero overhead if those fields are not present).
+
+
+## Implementation
+
+### Overview
+
+The implementation is broken into two parts:
+
+1. The `Buffer` which is almost identical to https://github.com/jhump/protoreflect/blob/master/dynamic/codec.go and provides an interface for iterating a marshaled protobuf message one `<fieldNumber, wiretype, value>` tuple at a time.
+
+2. The `customFieldUnmarshaler` which wraps the `Buffer` and exposes an interface for efficiently unmarshaling top-level scalar fields with no allocations, as well as a fallback mechanism that relies on the `jhump/protoreflect` library for any fields that don't have an efficient unmarshaling codepath (`maps`, `repeated` fields, and nested messages, etc.)
+
+### Buffer
+
+The code in the `Buffer` is mostly self explanatory for anyone familiar with [proto3 encoding format](https://developers.google.com/protocol-buffers/docs/encoding).
+
+### CustomFieldUnmarshaler
+
+The `customFieldUnmarshaler` has three primary responsibilities:
+
+1. Provide an interface for efficiently unmarshaling top-level scalar fields in a marshaled Protobuf message without allocating.
+2. Ensure that the values unmarshaled in #1 are sorted by field number.
+3. Provide a `*dynamic.Message` that contains *only* the fields that could not be unmarshaled efficiently in #1. This does not allocate / expend any resources at all in the case of optimized schemas that only contain fields that can be handled by #1.
+
+The `customFieldUnmarshaler` works by iterating through all the `<fieldNumber, wireType, value>` tuples in the marshaled Protobuf and checking if they are supported by the efficient code path.
+If they are, then it unmarshals the value into an `unmarshalValue` which is a space-optimized type that can be re-used without any allocations.
+If the tuple cannot be unmarshaled efficiently then the unmarshaler falls back to the `jhump/protoreflect` library and uses the `UnmarshalMerge` method to unmarshal a single field at a time so that it can be used iteratively without having to rewrite the marshaled message even if custom and non-custom fields are interleaved together.
```

Comment:
[conditional sentences](https://www.grammarly.com/blog/conditional-sentences/) should generally be comma-delimited:

> [...] efficiently, the marshaler [...]

minor tweak for readability (dang, that's a long sentence):

> [...] library and uses `UnmarshalMerge` to iteratively unmarshal each field sequentially to avoid rewriting the message. This approach supports both any combination of custom and non-custom fields, and in any order.